# 1. Data Validation
## Product Sales Strategy Analysis

The purpose of this notebook is to validate the quality, structure, and
integrity of the product sales dataset prior to analysis.

This phase focuses on identifying data issues such as missing values,
inconsistent categories, unrealistic ranges, and structural limitations.
No analytical conclusions are drawn at this stage.


## Introduction & Business Objective:

The business recently launched a new line of office stationery designed to encourage creativity and improve brainstorming effectiveness. To support the sales team, an analysis was conducted to evaluate the effectiveness of three sales strategies used to promote this product line: email outreach, phone calls, and a combined email-and-call approach.

The objective of this analysis is to compare the performance of these sales strategies, identify which approach leads to better customer engagement and sales outcomes, and provide data-driven recommendations to improve future sales efforts.

## Dataset Overview:

The dataset contains information on customer outreach attempts and sales outcomes for the new office stationery product line. Each row represents a sales interaction, including the sales strategy used, customer engagement indicators, and purchase outcomes.

Key variables include:
- Sales strategy used (email, call, or combination)
- Customer response indicators
- Sales or converson outcomes

This dataset is used to evaluate how different sales approaches influence customer behavior and overall sales performance.

## Import Libraries:

In [31]:
# Libraries:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

## Ingest Dataset(csv):

In [6]:
# Ingest Raw data:

p_sales = pd.read_csv("../data/raw/product_sales.csv")

## Data Validation:

In [27]:
# Data Validation:

print(p_sales.shape)
print(p_sales.shape)
print(p_sales.columns)
p_sales.info()
p_sales.head()

(15000, 8)
(15000, 8)
Index(['week', 'sales_method', 'customer_id', 'nb_sold', 'revenue',
       'years_as_customer', 'nb_site_visits', 'state'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   week               15000 non-null  int64  
 1   sales_method       15000 non-null  object 
 2   customer_id        15000 non-null  object 
 3   nb_sold            15000 non-null  int64  
 4   revenue            13926 non-null  float64
 5   years_as_customer  15000 non-null  int64  
 6   nb_site_visits     15000 non-null  int64  
 7   state              15000 non-null  object 
dtypes: float64(1), int64(4), object(3)
memory usage: 937.6+ KB


week  sales_method                           customer_id  nb_sold  revenue  \
0     2         Email  2e72d641-95ac-497b-bbf8-4861764a7097       10      NaN   
1     6  Email + Call  3998a98d-70f5-44f7-942e-789bb8ad2fe7       15   225.47   
2     5          Call  d1de9884-8059-4065-b10f-86eef57e4a44       11    52.55   
3     4         Email  78aa75a4-ffeb-4817-b1d0-2f030783c5d7       11      NaN   
4     3         Email  10e6d446-10a5-42e5-8210-1b5438f70922        9    90.49   

   years_as_customer  nb_site_visits      state  
0                  0              24    Arizona  
1                  1              28     Kansas  
2                  6              26  Wisconsin  
3                  3              25    Indiana  
4                  0              28   Illinois

In [26]:
# check for multi transactions:

multiple_transactions = p_sales["customer_id"].value_counts().max() > 1
print(multiple_transactions)

False


In [28]:
# missing values:

p_sales.isna().sum()

# we have 1074 missing values for revenue

week                    0
sales_method            0
customer_id             0
nb_sold                 0
revenue              1074
years_as_customer       0
nb_site_visits          0
state                   0
dtype: int64

In [29]:
# looking for unique values:

p_sales.nunique()

week                     6
sales_method             5
customer_id          15000
nb_sold                 10
revenue               6743
years_as_customer       42
nb_site_visits          27
state                   50
dtype: int64

In [25]:
# Desciptive stats:

p_sales.describe()

week       nb_sold       revenue  years_as_customer  \
count  15000.000000  15000.000000  13926.000000       15000.000000   
mean       3.098267     10.084667     93.934943           4.965933   
std        1.656420      1.812213     47.435312           5.044952   
min        1.000000      7.000000     32.540000           0.000000   
25%        2.000000      9.000000     52.470000           1.000000   
50%        3.000000     10.000000     89.500000           3.000000   
75%        5.000000     11.000000    107.327500           7.000000   
max        6.000000     16.000000    238.320000          63.000000   

       nb_site_visits  
count    15000.000000  
mean        24.990867  
std          3.500914  
min         12.000000  
25%         23.000000  
50%         25.000000  
75%         27.000000  
max         41.000000

In [30]:
# validate sales_method feature:

p_sales["sales_method"].unique()

array(['Email', 'Email + Call', 'Call', 'em + call', 'email'],
      dtype=object)

### Data Validation Summary:
positively skewd (or right skewed) distribution means we have a longer tail that extends in the positive direction (right)Dataset Structure & Integrity Checks:

The product sales dataset consists of 15,000 observations and 8 features, with one row representing one customer. This structure was validated by confirming that the customer_id field contains 15,000 unique values, which aligns with the expectation of one record per customer

The dataset includes the following key variables:
- sales method
- number of years a custimer has been
- number sold
- revenue
- number of site visits
- state location of customer
  
All columns were reviewed for data type correctness and completeness.
 
Structures:
- Rows: 15000 observations
- Columns: 8 features
- One row per customer
- so far datatypes are set accordingly

key fields:
- week: counted as weeks since product launch
- sales_method: three sales methods were used for that customer
- customer_id: unique identifier for the customer
- nb_sold: number of new products sold
- revenue: revenue from the sales, rounded to 2 decimal places
- years_as_customer: umber of years customer has been buying from us (company founded in 1984)
- nb_site_visits: number of times the customer has visited our website in the last 6 months
- state: location of the customer i.e. where orders are shipped

Sales Method Validation:

Initial validation of the sales_method column revealed five unique values, despite the business using only three intended sales strategies

Observed values:
- Email
- email
- Email + Call
- em + call

This inconsistency was caused by case sensitivity and typographical variations. These values represent the same underlying
strategies and were therefore standardized into three categories:
- mail
- Call
- Email + Call

This cleaning step ensures accurate grouping and comparison during analysis.

notes:
- revenue column has missing values (only 13,926 entries out of 15,000 are present). All other numerical columns are complete
- data covers a short sales period, spanning only 6 weeks (min 1, max 6)
- on avg customer buys about 10 products with most sales falling within range of 9 to 11
- i noticed that customers a quite active on the website with consistent traffic between 23 and 27 visits for the middle 50% of customers
- customer_id has high cardinality of 15000 which is expected confirming the one row per customer observation
- revenue has high cardinality values with a total of 6743 unique values
- years_as_customer pecullarly have an etry of max: 63 which supposedly till 40year since the company was build in 1984 only
- revenue mean is slightly higher with 93.93 vs median of 89.50 indicating a slightly positive skew
- positively skewd (or right skewed) distribution means we have a longer tail that extends in the positive direction (right)